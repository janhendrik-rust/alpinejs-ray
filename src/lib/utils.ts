/* eslint-disable no-undef */

export function getWindow() {
    // @ts-ignore
    return <any>globalThis;
}

/**
 * Determines if the axios library has been installed and initialized in the current browser environment.
 *
 * @param window
 */
export const checkForAxios = (window: any = null) => {
    const win: any = window ?? getWindow();

    if (!win.axios) {
        throw new Error('[alpinejs-ray] axios is required for alpinejs-ray to function correctly.');
    }
};

/**
 * Determines if the node-ray standalone library has been installed and initialized in the current browser environment.
 *
 * @param window
 */
export const checkForRay = (window: any = null) => {
    const win: any = window ?? getWindow();

    if (!win.Ray || !win.Ray.Ray || !win.Ray.ray) {
        throw new Error('[alpinejs-ray] node-ray is required for alpinejs-ray to function correctly.');
    }
};

/**
 * Determines if alpine.js has been installed and initialized in the current browser environment.
 *
 * @param window
 */
export const checkForAlpine = (window: any = null) => {
    const win: any = window ?? getWindow();

    if (!win.Alpine) {
        throw new Error('[alpinejs-ray] Alpine is required for alpinejs-ray to function correctly.');
    }

    if (!win.Alpine.version || !isValidVersion('2.5.0', win.Alpine.version)) {
        throw new Error('Invalid Alpine version. Please use Alpine version 2.5.0 or above');
    }
};

/**
 * returns true if the `current` version string is greater than or equal to `required`
 * version string.
 *
 * this converts version strings like '2.5.1' to 100210051001, then compares values.
 *
 * @param required string
 * @param current string
 * @returns boolean
 */
export function isValidVersion(required: string, current: string) {
    let currentVersionIdentStr: string = '',
        requiredVersionIdentStr: string = '';

    let requiredArray: number[] = required.split('.').map(part => 1000 + parseInt(part));
    let currentArray: number[] = current.split('.').map(part => 1000 + parseInt(part));

    while (currentArray.length < requiredArray.length) {
        currentArray.push(1000);
    }

    // ensure the array lengths match
    currentArray = currentArray.slice(0, requiredArray.length - 1);

    for (let idx = 0; idx < currentArray.length; idx++) {
        currentVersionIdentStr += currentArray[idx].toString();
        requiredVersionIdentStr += requiredArray[idx].toString();
    }

    return Number(currentVersionIdentStr) >= Number(requiredVersionIdentStr);
}

export const encodeHtmlEntities = (str: string) => {
    const escapeChars: Record<string, string> = {
        '¢': 'cent',
        '£': 'pound',
        '¥': 'yen',
        '€': 'euro',
        '©': 'copy',
        '®': 'reg',
        '<': 'lt',
        '>': 'gt',
        '"': 'quot',
        '&': 'amp',
        "'": '#39',
        ' ': 'nbsp',
    };

    const chars: string[] = Object.keys(escapeChars);
    const regex = new RegExp(`[${chars.join('')}]`, 'g');

    return str.replace(regex, m => `&${escapeChars[m]};`);
};

export const filterObjectKeys = (obj: Record<string, unknown>, ignoreKeys: string[]): Record<string, unknown> => {
    const result = {};

    Object.keys(obj)
        .filter(key => !ignoreKeys.includes(key))
        .forEach(key => {
            result[key] = obj[key];
        });

    return result;
};

export const highlightHtmlMarkup = (str: string) => {
    return encodeHtmlEntities(str)
        .replace(/(:?href|:?id|:?name|:?value|:?type|:?style|:?class|:?key)=/g, '<span class="text-orange-400">$1</span>=')
        .replace(/&#39;(.+)&#39;/g, '<span class="text-purple-600">&#39;$1&#39;</span>')
        .replace(/&quot;(.*)&quot;/g, '<span class="text-green-600">&quot;$1&quot;</span>')
        .replace(/\r?\n/g, '<br>')
        .replace(/(&lt;)(\/?\s*[\w-_]+)(&gt;)?/g, '<span class="text-blue-600">$1$2$3</span>')
        .replace(
            /(@click|x-bind|x-cloak|x-data|x-for|x-if|x-html|x-init|x-model[\.\-\w+]*|x-on:\w+[\.\-\w+]*|x-ref|x-show|x-spread|x-transition:\w+[\-\w]+|x-text)=/g,
            '<span class="text-indigo-800">$1</span>='
        )
        .replace(/(\$el|\$refs|\$event|\$data|\$dispatch|\$nextTick|\$watch|\$ray)/g, '<span class="text-red-700">$1</span>');
};
