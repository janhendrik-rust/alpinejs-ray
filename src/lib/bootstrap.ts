interface BootstrapParameters {
    AlpineRayPlugin: any;
    ray: any;
    Ray: any;
    axios: any;
    Alpine: any;
}

import { getWindow } from './utils';

export const bootstrap = (params: BootstrapParameters, window: any = null) => {
    //({ plugin, ray, Ray, axios, Alpine, window }) => {
    window = window ?? getWindow();

    for (const prop in params) {
        window[prop] = params[prop];
    }

    window.axios.defaults.headers.common['X-Requested-With'] = 'XMLHttpRequest';

    window.AlpineRayPlugin.init();

    return window.AlpineRayPlugin;
};
