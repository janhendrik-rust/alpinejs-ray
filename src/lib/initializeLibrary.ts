import { ray } from '../AlpineRay';
import { addErrorEventHandlers } from '../ErrorHandlers';
import { getWindow } from '../lib/utils';
import { SpruceProxy } from '../SpruceProxy';

export const initializeLibrary = (plugins: any[], window: any = null, rayInstance: any = null) => {
    window = window ?? getWindow();
    rayInstance = rayInstance ?? ray();

    initializePlugins(plugins, window, rayInstance);
    initializeErrorEventHandlers(window, rayInstance);
    initializeSpruce(window, rayInstance);
};

export const initializePlugins = (plugins: any[], window: any = null, rayInstance: any = null) => {
    window = window ?? getWindow();
    rayInstance = rayInstance ?? ray();

    plugins.forEach(am => {
        am.init(window, rayInstance);
    });
};

export const initializeErrorEventHandlers = (window: any = null, rayInstance: any = null) => {
    window = window ?? getWindow();
    rayInstance = rayInstance ?? ray();

    addErrorEventHandlers(window, rayInstance);
};

export const initializeSpruce = (window: any = null, rayInstance: any = null) => {
    window = window ?? getWindow();
    rayInstance = rayInstance ?? ray();

    // conditional Spruce proxy/monitor
    new SpruceProxy(window, rayInstance).init();
};
