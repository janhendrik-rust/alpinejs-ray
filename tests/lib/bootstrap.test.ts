/* eslint-disable no-undef */
import { bootstrap } from '../../src/lib/bootstrap';

it('bootstraps alpinejs-ray imports for a bundler', () => {
    let win = {};
    const Alpine = 'AlpineClass';
    const Ray = 'RayClass';
    const ray = 'rayHelper';
    const axios = {
        defaults: {
            headers: {
                common: {},
            },
        },
    };
    const AlpineRayPlugin = {
        initRunCount: 0,
        init() {
            this.initRunCount += 1;
        },
    };

    const result: any = bootstrap({ Alpine, AlpineRayPlugin, Ray, ray, axios }, win);

    expect(AlpineRayPlugin.initRunCount).toBe(1);
    expect(result).toBe(AlpineRayPlugin);
    expect(win).toMatchSnapshot();
});
