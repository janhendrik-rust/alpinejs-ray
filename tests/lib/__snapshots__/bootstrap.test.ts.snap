// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`bootstraps alpinejs-ray imports for a bundler 1`] = `
Object {
  "Alpine": "AlpineClass",
  "AlpineRayPlugin": Object {
    "init": [Function],
    "initRunCount": 1,
  },
  "Ray": "RayClass",
  "axios": Object {
    "defaults": Object {
      "headers": Object {
        "common": Object {
          "X-Requested-With": "XMLHttpRequest",
        },
      },
    },
  },
  "ray": "rayHelper",
}
`;
